### PR TITLE
A few fixes related to CommonJS module loading

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

- Add extra checks to ensure source code is always freed when it should be
- Optimization: precompute the hash upfront for source code transpiled on another thread. This seems
- Make `Bun.gc(true)` free more memory
- Make `CommonJSModule` store the JSC::SourceCode directly instead of using `JSC::JSSourceCode`. Goal is to free it earlier instead of whenever GC decides to free it.
- Fixes https://github.com/oven-sh/bun/issues/6946
- Fixes https://github.com/oven-sh/bun/issues/8965
- Fixes a bug where bun loads & transpiles the same code twice(!)

### How did you verify your code works?

It doesn't work yet 

It crashes in release builds, but not debug builds

I suspect something fishy with either GC or a C++ use-after-move somewhere. Worst case I revert it back

### CODE REVIEW

File: src/bun.js/module_loader.zig [hunk 1]
```diff
@@ -637,20 +665,31 @@ pub const RuntimeTranspilerStore = struct {
...
-                    break :brk result;
-                },
+                // In a benchmarking loading @babel/standalone 100 times:
+                //
+                // Before ensureHash:
+                // 354.00 ms    4.2%	354.00 ms	 	  WTF::StringImpl::hashSlowCase() const
+                //
+                // After ensureHash:
+                // 506.00 ms    6.1%	506.00 ms	 	  WTF::StringImpl::hashSlowCase() const
+                //
```
Comment (paperclover) [inline src/bun.js/module_loader.zig:685]: is this in the right order

**Comment by staabm** [comment]
> * Optimization: precompute the hash upfront for source code transpiled on another thread. This seems

nit: it seems the PR description is truncated at this point

**Review by paperclover — APPROVED** [review]
looks good now.

if you are worried about JS code being wrong, feel free to add more `$assert`s into it (they are tree-shaken)

File: src/bun.js/module_loader.zig (hunk 1 above)
Comment (Jarred-Sumner) [reply src/bun.js/module_loader.zig:685]: image

**Pull request merged**